### PR TITLE
RKUsbLoader/RKUSBDriver and minor fixes

### PR DESCRIPTION
**Pull request by otavio**
**Description**

**Checklist**
- [x] Documentation for the feature
- [ ] Tests for the feature 

- [x] The arguments and description in doc/configuration.rst have been updated

- [x] Add a section on how to use the feature to doc/usage.rst

- [x] Add a section on how to use the feature to doc/development.rst

- [ ] CHANGES.rst has been updated

- [x] PR has been tested

- [ ] Man pages have been regenerated

### CODE REVIEW

File: labgrid/driver/usbloader.py [hunk 1]
```diff
@@ -82,3 +83,54 @@ def load(self, filename=None):
...
+        pass
+
+    def on_deactivate(self):
+        pass
+
+    @Driver.check_active
+    @step(args=['filename'])
+    def load(self, filename=None):
+        # FIXME make sure we always have an environment or config
+        loaders = self.target.env.config.get_tool('loader') or 'loader'
```
Comment (jluebbe) [inline labgrid/driver/usbloader.py]: Loader seems to be a very generic name. Is this an additional image which needs to be loaded before the actual bootloader? If so, it should probably be set via an attribute on the driver (like the image itself).

File: labgrid/driver/usbloader.py [hunk 2]
```diff
@@ -82,3 +83,54 @@ def load(self, filename=None):
...
+    def on_activate(self):
+        pass
+
+    def on_deactivate(self):
+        pass
+
+    @Driver.check_active
+    @step(args=['filename'])
+    def load(self, filename=None):
+        # FIXME make sure we always have an environment or config
```
Comment (jluebbe) [inline labgrid/driver/usbloader.py]: Yes, you need to check if you have an environment set (usally ``if self.target.env``).

File: labgrid/driver/usbloader.py [hunk 3]
```diff
@@ -82,3 +83,55 @@ def load(self, filename=None):
...
+            + [self.tool, 'db', mf.get_remote_path()]
+        )
+
+        if filename is None and self.image is not None:
+            filename = self.target.env.config.get_image_path(self.image)
+        mf = ManagedFile(filename, self.loader)
+        mf.sync_to_resource()
+
+        # wait the loader finish the process
+        time.sleep(1)
```
Comment (Emantor) [inline labgrid/driver/usbloader.py]: Why is this sleep necessary here?

Comment (otavio) [reply labgrid/driver/usbloader.py]: We can confirm if the `sleep` is required. Worse, we can do a busy loop, trying to load it with a max of 3s for timeout.

**Comment by chtavares592** [comment]
@Emantor I need to use something that wait for the loader image start at the board. So how @otavio told you,  I edited the code and now it does a loop, trying to load it with a max of 3s for the timeout.

File: labgrid/driver/usbloader.py [hunk 4]
```diff
@@ -78,7 +79,61 @@ def load(self, filename=None):
...
+
+        if filename is None and self.image is not None:
+            filename = self.target.env.config.get_image_path(self.image)
+        mf = ManagedFile(filename, self.loader)
+        mf.sync_to_resource()
+
+        for _ in range(30):
+           # The u-boot image is written on the device
+            try:
+                time.sleep(0.1)
```
Comment (Emantor) [inline labgrid/driver/usbloader.py]: I prefer something along the lines of
```
deadline = time.monotonic() + 3.0
while time.monotonic() < deadline:
```
as a timeout construct.

Comment (chtavares592) [reply labgrid/driver/usbloader.py]: @Emantor modifications were performed

**Comment by otavio** [comment]
Hello; what is pending for this to move on? @Emantor any other needed change?

**Comment by chtavares592** [comment]
@jluebbe I did rebase on a master and did push, now it's all right.

File: labgrid/driver/usbloader.py [hunk 5]
```diff
@@ -79,6 +80,66 @@ def load(self, filename=None):
...
+        deadline = time.monotonic() + 3.0
+        while time.monotonic() < deadline:
+            try:
+                subprocess.check_call(
+                    self.loader.command_prefix +
+                    [self.tool, 'db', mf.get_remote_path()]
+                )
+                break
+            except subprocess.CalledProcessError:
+                continue
```
Comment (jluebbe) [inline labgrid/driver/usbloader.py]: If this times out, we will just continue with the next stage, without raising the exception.
With Timeout from labgrid.util, you could do:
```python
    timeout = Timeout(3.0)
    while True:
        try:
            subprocess.check_call(
                self.loader.command_prefix +
                [self.tool, 'db', mf.get_remote_path()]
            )
            break
        except subprocess.CalledProcessError:
            if timeout.expired:
                raise
```

File: labgrid/driver/usbloader.py [hunk 6]
```diff
@@ -79,6 +80,66 @@ def load(self, filename=None):
...
+        deadline = time.monotonic() + 3.0
+        while time.monotonic() < deadline:
+            try:
+                subprocess.check_call(
+                    self.loader.command_prefix
+                    + [self.tool, 'wl', '0x40', mf.get_remote_path()]
+                )
+                break
+            except subprocess.CalledProcessError:
+                continue
```
Comment (jluebbe) [inline labgrid/driver/usbloader.py]: This has the same issue as above.

**Comment by chtavares592** [comment]
@jluebbe the changes have been applied

**Comment by jluebbe** [comment]
@otavio Looking at this after some time, I noticed that your `rk-usb-loader` doesn't seem to be available? Is that a wrapper/fork of https://github.com/rockchip-linux/rkdeveloptool?

**Comment by otavio** [comment]
Yes. It is this tool.

In our labgrid config we use:
```yaml
targets:
  main:
    resources:
      RKUSBLoader:
        match:
          'sys_name': '1-1.1.3'
      USBMassStorage:
        match:
          'ID_PATH': 'platform-3f980000.usb-usb-0:1.1.3:1.0-scsi-0:0:0:0'
      USBSerialPort:
        match:
          'ID_PATH': 'platform-3f980000.usb-usb-0:1.3:1.0'
        speed: 1500000
    drivers:
      RKUSBDriver:
        image: bootloader
        usb_loader: boot-rom

tools:
  rk-usb-loader: /usr/bin/rkdeveloptool
images:
  bootloader: images/u-boot.img
  boot-rom: images/RV1108_usb_boot.bin
```

**Comment by otavio** [comment]
Please do whatever is need. We can update here.

**Pull request merged**